### PR TITLE
Suppression state propagation

### PR DESCRIPTION
**Pull request by SandyRogers**
This PR:
- adds a "dependency tree" to propagate suppression states between models.
  - e.g. when a study is suppressed, the samples, runs, assemblies, analyses that belong to that study are also suppressed
  - the complexity of this is really so that it can be done in bulk operations. It could be simplified if we were happy to have slow jobs that update objects one by one.
- adds a suppression state of "Ancestor Suppressed" to keep track of these events.
  - e.g. so that a descendant (say, a sample) can remain suppressed even if its ancestor (say, a study) is unsuppressed – only descendants with the "Ancestor Suppressed" state should be unsuppressed.
- adds a `select_related_only` queryset method, to reduce the amount of data selected by complex queries like the `AnalysisJob`. 
  - This lets us keep necessary joins (like getting the `study.accession` for an `analysisjob`), without querying the many other fields on `sample` that are never picked by the serializer.

Together these simplify the querying of some models, because suppression states of related objects do not have to be checked at query time and a slightly smaller query is used. This has been implemented on the Analysis endpoint currently, and seems to give a speedup of around 20%... to get more improvement, we need to do something similar with the `is_private` states. These are more complicated because we need to add `submission_account` info on other models, and check this depending on the request user. I have not done that in this PR, but it might be worth a go in future...

### CODE REVIEW

File: emgapi/models.py
Comment (mberacochea) [inline emgapi/models.py:120]:
```suggestion
            # (If so, it shouldn't be suppressed).
            # This was written mostly for samples that are associated to multiples 
            # studies, such as a raw-reads study and the corresponding assembly study.
```

File: emgapi/models.py
Comment (mberacochea) [inline emgapi/models.py:210]: I'm curious... why "defer" and not just use [`.only()`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#only). Using defer is more flexible as the data will be there, but it may introduce may cause the ORM to make more queries (if my understanding is correct)

**Review by mberacochea — APPROVED** [review]
This is great stuff @SandyRogers . I've added 2 minor comments. I think @MGS-sails should give this PR a look before merging.

File: emgapi/models.py
Comment (SandyRogers) [reply emgapi/models.py:210]: This is because you can't apply an `.only` just to the fields of a `.select_related`, e.g. you would chain them like:
```python
>>> Run.objects.select_related('sample').only('sample__sample_id')
Traceback (most recent call last):
...
django.core.exceptions.FieldError: Field Run.study cannot be both deferred and traversed using select_related at the same time.
```

(I.e. you'd have to re-add all the fields of the local side of the related to the `only(...)`)

There is probably a way involving subqueries or annotations I suppose, but this use of defer to create a sort of related-scoped `only` seemed a bit easier to reckon with.

If used carefully then no extra queries should be created in the places we use this – it is designed for use in queries where we just need a couple of fields like Accession from the other side of relationship, to render a minimal data into the hyperlinked related relationships.

File: emgapi/models.py
Comment (MGS-sails) [inline emgapi/models.py:192]: @SandyRogers this might be overkill, and please ignore if it's going to cause an uneccessary delay, 
but I think in the long term, it will be better to refactor the suppress and unsuppress methods.
It seems like they are doing the same thing, with only slight variations. 
So it might be better if they both call a generalised method, but with different arguments.

**Review by MGS-sails — COMMENTED** [review]
Left a couple of comments. 
Thanks

File: emgapi/models.py
Comment (mberacochea) [reply emgapi/models.py:212]: perfect.. I wasn't aware of that caveat of "only()". Thanks :)

**Comment by mberacochea** [comment]
@SandyRogers should we merge this one?

**Pull request merged**
